### PR TITLE
fix(tianmu): the EXTRACT() function error.(#845)

### LINKED ISSUE #845: bug: the EXTRACT() function error
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

The result returned by the function EXTRACT() is incorrect
The wrong result is as follows:
```shell
mysql> select * from ctimestamp;
+---------------------+
| a                   |
+---------------------+
| 2019-01-02 00:02:03 |
| 2019-01-02 01:02:03 |
| 2019-01-02 10:11:12 |
+---------------------+
3 rows in set (0.00 sec)

mysql> select EXTRACT(DAY_HOUR FROM a), EXTRACT(MINUTE_SECOND FROM a) from ctimestamp;
+--------------------------+-------------------------------+
| EXTRACT(DAY_HOUR FROM a) | EXTRACT(MINUTE_SECOND FROM a) |
+--------------------------+-------------------------------+
|                      200 |                           200 |
|                      201 |                           201 |
|                      210 |                           210 |
+--------------------------+-------------------------------+
3 rows in set (0.01 sec)
```

### Expected behavior

```shell
mysql> select EXTRACT(DAY_HOUR FROM a), EXTRACT(MINUTE_SECOND FROM a) from ctimestamp;
+--------------------------+-------------------------------+
| EXTRACT(DAY_HOUR FROM a) | EXTRACT(MINUTE_SECOND FROM a) |
+--------------------------+-------------------------------+
|                      200 |                           203 |
|                      201 |                           203 |
|                      210 |                          1112 |
+--------------------------+-------------------------------+
3 rows in set (0.00 sec)
```

### How To Reproduce

```sql
CREATE TABLE ctimestamp (a timestamp);
INSERT INTO ctimestamp VALUES ('2019-01-02 00:02:03'),
('2019-01-02 01:02:03'), ('2019-01-02 10:11:12');
select EXTRACT(DAY_HOUR FROM a), EXTRACT(MINUTE_SECOND FROM a) from ctimestamp;
```

### Environment

```shell
root@ub01:~# /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
        Branch name: stonedb-5.7-dev
        Last commit ID: 86a7685dc
        Last commit time: Date:   Sat Oct 22 03:38:32 2022 +0800
        Build time: Date: Sat 22 Oct 2022 02:14:05 PM UTC
root@ub01:~# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by DandreChen**
## Summary about this PR

Issue Number: close #845 

[summary]
add item_name comparison for the diff operation of the same function,as EXTRACT(DAY_HOUR FROM t) and EXTRACT(MINUTE_SECOND FROM t).
fix some warnings

## Tests Check List

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Review by lujiashun — APPROVED** [review]
LGTM

File: storage/tianmu/core/mysql_expression.cpp [hunk 1]
```diff
@@ -618,22 +621,14 @@ bool operator==(Item const &l_, Item const &r_) {
...
           if (same && l->const_item())
             same = ((Item_func *)&l_)->val_int() == ((Item_func *)&r_)->val_int();
-          if (dynamic_cast<const Item_date_add_interval *>(&l_)) {
-            const Item_date_add_interval *l = static_cast<const Item_date_add_interval *>(&l_);
-            const Item_date_add_interval *r = static_cast<const Item_date_add_interval *>(&r_);
-            same = same && dynamic_cast<const Item_date_add_interval *>(&r_);
-            same = same && ((l->int_type == r->int_type) && (l->date_sub_interval == r->date_sub_interval));
-          }
+
+          const Item_date_add_interval *l = static_cast<const Item_date_add_interval *>(&l_);
```
Comment (RingsC) [inline storage/tianmu/core/mysql_expression.cpp]: uses `down_cast` not `static_cast` directly.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/mysql_expression.cpp (hunk 1 above)
Comment (DandreChen) [reply storage/tianmu/core/mysql_expression.cpp]: fixed it

Comment (DandreChen) [reply storage/tianmu/core/mysql_expression.cpp]: there can not use down_cast,  will lead to some mtr cases crash  in debug mode,  as below case:
```
CREATE TABLE `select_base_with_primary_key` (
  `pk` bigint(11) NOT NULL AUTO_INCREMENT,
  `varchar_test` varchar(255) DEFAULT NULL,
  `integer_test` int(11) DEFAULT NULL,
  `char_test` char(255) DEFAULT NULL,
  `tinyint_test` tinyint(4) DEFAULT NULL,
  `tinyint_1bit_test` tinyint(1) DEFAULT NULL,
  `smallint_test` smallint(6) DEFAULT NULL,
  `mediumint_test` mediumint(9) DEFAULT NULL,
  `bigint_test` bigint(20) DEFAULT NULL,
  `double_test` double DEFAULT NULL,
  `decimal_test` decimal(10,0) DEFAULT NULL,
  `date_test` date DEFAULT NULL,
  `time_test` time DEFAULT NULL,
  `datetime_test` datetime DEFAULT NULL,
  `timestamp_test` timestamp NULL DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
  `year_test` year(4) DEFAULT NULL,
  PRIMARY KEY (`pk`)
) ENGINE=TIANMU DEFAULT CHARSET=gbk;
insert into select_base_with_primary_key values (11, 'abcde', 123, 'char', 3, 1, 23, 32, 332, 1.4, 25, '2017-11-11', '11:11:11', '2017-11-11 11:11:11','2017-11-11 11:11:11', 2017);
select 'aa' as ag_column1,  !( tinyint_1bit_test )as ag_column2, 1 as ag_column3,min((smallint_test +  smallint_test) =11)as ag_column_fun  from select_base_with_primary_key    GROUP BY  ag_column1,ag_column2  asc,ag_column3   having ag_column_fun<= 37 order by  ag_column1    ,ag_column2    asc,ag_column3    asc,ag_column_fun;
```
stack：
```
down_cast<Item_date_add_interval const*, Item const*>(const Item * arg) (\opt\github\stonedb57\include\template_utils.h:85)
Tianmu::core::operator==(const Item & l_, const Item & r_) (\opt\github\stonedb57\storage\tianmu\core\mysql_expression.cpp:628)
Tianmu::core::MysqlExpression::operator==(const Tianmu::core::MysqlExpression * const this, const Tianmu::core::MysqlExpression & other) (\opt\github\stonedb57\storage\tianmu\core\mysql_expression.cpp:684)
Tianmu::core::Query::VirtualColumnAlreadyExists(Tianmu::core::Query * const this, const Tianmu::core::TabID & tmp_table, Tianmu::core::MysqlExpression * expression) (\opt\github\stonedb57\storage\tianmu\core\query.cpp:139)
Tianmu::core::Query::AddColumnForMysqlExpression(Tianmu::core::Query * const this, Tianmu::core::MysqlExpression * mysql_expression, const Tianmu::core::TabID & tmp_table, const char * alias, const Tianmu::common::ColOperation oper, const bool distinct, bool group_by) (\opt\github\stonedb57\storage\tianmu\core\query_compile.cpp:849)
Tianmu::core::Query::AddFields(Tianmu::core::Query * const this, List<Item> & fields, const Tianmu::core::TabID & tmp_table, const bool group_by_clause, int & num_of_added_fields, bool ignore_minmax, bool & aggregation_used) (\opt\github\stonedb57\storage\tianmu\core\query_compile.cpp:474)
Tianmu::core::Query::Compile(Tianmu::core::Query * const this, Tianmu::core::CompiledQuery * compiled_query, SELECT_LEX * selects_list, SELECT_LEX * last_distinct, Tianmu::core::TabID * res_tab, bool ignore_limit, Item * left_expr_for_subselect, Tianmu::common::Operator * oper_for_subselect, bool ignore_minmax, bool for_subq_in_where) (\opt\github\stonedb57\storage\tianmu\core\query_compile.cpp:1112)
Tianmu::core::Engine::Execute(Tianmu::core::Engine * const this, THD * thd, LEX * lex, Query_result * result_output, SELECT_LEX_UNIT * unit_for_union) (\opt\github\stonedb57\storage\tianmu\core\engine_execute.cpp:436)
Tianmu::core::Engine::HandleSelect(Tianmu::core::Engine * const this, THD * thd, LEX * lex, Query_result *& result, ulong setup_tables_done_option, int & res, int & optimize_after_tianmu, int & tianmu_free_join, int with_insert) (\opt\github\stonedb57\storage\tianmu\core\engine_execute.cpp:238)
Tianmu::handler::ha_my_tianmu_query(THD * thd, LEX * lex, Query_result *& result_output, ulong setup_tables_done_option, int & res, int & optimize_after_tianmu, int & tianmu_free_join, int with_insert) (\opt\github\stonedb57\storage\tianmu\handler\ha_my_tianmu.cpp:88)
execute_sqlcom_select(THD * thd, TABLE_LIST * all_tables) (\opt\github\stonedb57\sql\sql_parse.cc:5184)
mysql_execute_command(THD * thd, bool first_level) (\opt\github\stonedb57\sql\sql_parse.cc:2831)
mysql_parse(THD * thd, Parser_state * parser_state) (\opt\github\stonedb57\sql\sql_parse.cc:5622)
dispatch_command(THD * thd, const COM_DATA * com_data, enum_server_command command) (\opt\github\stonedb57\sql\sql_parse.cc:1495)
do_command(THD * thd) (\opt\github\stonedb57\sql\sql_parse.cc:1034)
handle_connection(void * arg) (\opt\github\stonedb57\sql\conn_handler\connection_handler_per_thread.cc:313)
```
breakpiont
```
template<typename Target, typename Source>
inline Target down_cast(Source arg)
{
  assert(NULL != dynamic_cast<Target>(arg));
  return static_cast<Target>(arg);
}

```

Comment (DandreChen) [reply storage/tianmu/core/mysql_expression.cpp]: `l_` real type is:`Item_func_not`
```
$2 = (const Item_func_not &) @0x7ffc3c010b50: {
  <Item_bool_func> = {
    <Item_int_func> = {
      <Item_func> = {
        <Item_result_field> = {
          <Item> = {
            <Parse_tree_node> = {
              _vptr.Parse_tree_node = 0x4025398 <vtable for Item_func_not+16>,
              contextualized = true,
              transitional = false
            },
            members of Item:
            is_expensive_cache = -1 '\377',
            rsize = 0,
            str_value = {
              m_ptr = 0x0,
              m_length = 0,
              m_charset = 0x412e920 <my_charset_bin>,
              m_alloced_length = 0,
              m_is_alloced = false
            },
            item_name = {
              <Name_string> = {
                <Simple_cstring> = {
                  m_str = 0x7ffc3c010c28 "ag_column2",
                  m_length = 10
                }, <No data fields>},
              members of Item_name_string:
              m_is_autogenerated = false
--Type <RET> for more, q to quit, c to continue without paging--
            },
            orig_name = {
              <Name_string> = {
                <Simple_cstring> = {
                  m_str = 0x0,
                  m_length = 0
                }, <No data fields>},
              members of Item_name_string:
              m_is_autogenerated = true
            },
            next = 0x7ffc3c010a30,
            max_length = 1,
            marker = 0,
            decimals = 0 '\000',
            maybe_null = 1 '\001',
            null_value = 0 '\000',
            unsigned_flag = 0 '\000',
            with_sum_func = 0 '\000',
            fixed = 1 '\001',
            collation = {
              collation = 0x4136f20 <my_charset_latin1>,
              derivation = DERIVATION_NUMERIC,
              repertoire = 1
            },
            cmp_context = 4294967295,
            runtime_item = false,
            derived_used = false,
            with_subselect = 0 '\000',
            with_stored_program = 0 '\000',
--Type <RET> for more, q to quit, c to continue without paging--
            tables_locked_cache = false,
            is_parser_item = false
          },
          members of Item_result_field:
          result_field = 0x0
        },
        members of Item_func:
        args = 0x7ffc3c010bf8,
        tmp_arg = {0x7ffc3c901618, 0x0},
        const_item_cache = false,
        allowed_arg_cols = 1,
        used_tables_cache = 1,
        not_null_tables_cache = 1,
        arg_count = 1
      }, <No data fields>},
    members of Item_bool_func:
    m_created_by_in2exists = false
  }, <No data fields>}
```

`r_` real type is : `Item_func_eq`
```
$3 = (const Item_func_eq &) @0x7ffc3c010e88: {
  <Item_bool_rowready_func2> = {
    <Item_bool_func2> = {
      <Item_bool_func> = {
        <Item_int_func> = {
          <Item_func> = {
            <Item_result_field> = {
              <Item> = {
                <Parse_tree_node> = {
                  _vptr.Parse_tree_node = 0x40234c0 <vtable for Item_func_eq+16>,
                  contextualized = true,
                  transitional = false
                },
                members of Item:
                is_expensive_cache = -1 '\377',
                rsize = 0,
                str_value = {
                  m_ptr = 0x0,
                  m_length = 0,
                  m_charset = 0x412e920 <my_charset_bin>,
                  m_alloced_length = 0,
                  m_is_alloced = false
                },
                item_name = {
                  <Name_string> = {
                    <Simple_cstring> = {
                      m_str = 0x0,
                      m_length = 0
                    }, <No data fields>},
--Type <RET> for more, q to quit, c to continue without paging--
                  members of Item_name_string:
                  m_is_autogenerated = true
                },
                orig_name = {
                  <Name_string> = {
                    <Simple_cstring> = {
                      m_str = 0x0,
                      m_length = 0
                    }, <No data fields>},
                  members of Item_name_string:
                  m_is_autogenerated = true
                },
                next = 0x7ffc3c006c78,
                max_length = 1,
                marker = 0,
                decimals = 0 '\000',
                maybe_null = 1 '\001',
                null_value = 0 '\000',
                unsigned_flag = 0 '\000',
                with_sum_func = 0 '\000',
                fixed = 1 '\001',
                collation = {
                  collation = 0x4136f20 <my_charset_latin1>,
                  derivation = DERIVATION_NUMERIC,
                  repertoire = 1
                },
                cmp_context = 4294967295,
                runtime_item = false,
                derived_used = false,
--Type <RET> for more, q to quit, c to continue without paging--
                with_subselect = 0 '\000',
                with_stored_program = 0 '\000',
                tables_locked_cache = false,
                is_parser_item = false
              },
              members of Item_result_field:
              result_field = 0x0
            },
            members of Item_func:
            args = 0x7ffc3c010f30,
            tmp_arg = {0x7ffc3c006b98, 0x7ffc3c006c78},
            const_item_cache = false,
            allowed_arg_cols = 1,
            used_tables_cache = 1,
            not_null_tables_cache = 1,
            arg_count = 2
          }, <No data fields>},
        members of Item_bool_func:
        m_created_by_in2exists = false
      },
      members of Item_bool_func2:
      cmp = {
        <Sql_alloc> = {<No data fields>},
        members of Arg_comparator:
        a = 0x7ffc3c010f30,
        b = 0x7ffc3c010f38,
        func = (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da8fbe <Arg_comparator::compare_int_signed()>,
        owner = 0x7ffc3c010e88,
        comparators = 0x0,
--Type <RET> for more, q to quit, c to continue without paging--
        comparator_count = 0,
        precision = 0,
        a_cache = 0x0,
        b_cache = 0x0,
        is_nulls_eq = false,
        set_null = true,
        get_value_a_func = 0x0,
        get_value_b_func = 0x0,
        json_scalar = 0x0,
        cmp_collation = {
          collation = 0x412e920 <my_charset_bin>,
          derivation = DERIVATION_NONE,
          repertoire = 3
        },
        value1 = {
          m_ptr = 0x0,
          m_length = 0,
          m_charset = 0x412e920 <my_charset_bin>,
          m_alloced_length = 0,
          m_is_alloced = false
        },
        value2 = {
          m_ptr = 0x0,
          m_length = 0,
          m_charset = 0x412e920 <my_charset_bin>,
          m_alloced_length = 0,
          m_is_alloced = false
        },
        static comparator_matrix = {{
--Type <RET> for more, q to quit, c to continue without paging--
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da8488 <Arg_comparator::compare_string()>,
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da86c6 <Arg_comparator::compare_e_string()>}, {
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da885a <Arg_comparator::compare_real()>,
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da8b04 <Arg_comparator::compare_e_real()>}, {
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da8fbe <Arg_comparator::compare_int_signed()>,
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da952a <Arg_comparator::compare_e_int()>}, {
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da96d6 <Arg_comparator::compare_row()>,
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da98be <Arg_comparator::compare_e_row()>}, {
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da8970 <Arg_comparator::compare_decimal()>,
            (int (Arg_comparator::*)(Arg_comparator * const)) 0x1da8bee <Arg_comparator::compare_e_decimal()>}}
      },
      abort_on_null = false
    }, <No data fields>}, <No data fields>}
```

Comment (DandreChen) [reply storage/tianmu/core/mysql_expression.cpp]: `l` and `r ` type is `Item_date_add_interval` :
```
Item_date_add_interval -> Item_temporal_hybrid_func -> Item_temporal_hybrid_func
 -> Item_str_func -> Item_func -> Item_result_field->Item
```

Comment (DandreChen) [reply storage/tianmu/core/mysql_expression.cpp]: so ，cann't  use down_case

File: mysql-test/suite/tianmu/t/issue845.test [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+CREATE TABLE test_timestamp (t timestamp);
+INSERT INTO test_timestamp VALUES ('2019-01-02 00:01:02'),
+                              ('2019-01-02 01:02:03'),
+                              ('2019-01-02 02:03:04'),
+                              ('2019-01-02 03:05:06'),
+                              ('2019-01-02 04:07:08');
+
+#EXTRACT SELECT
+
+SELECT EXTRACT(DAY_HOUR FROM t), EXTRACT(MINUTE_SECOND FROM t) FROM test_timestamp;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue845.test:18]: adds some cases  just as following:
```
mysql> SELECT EXTRACT(YEAR FROM '2019-07-02');
        -> 2019
mysql> SELECT EXTRACT(YEAR_MONTH FROM '2019-07-02 01:02:03');
        -> 201907
mysql> SELECT EXTRACT(DAY_MINUTE FROM '2019-07-02 01:02:03');
        -> 20102
mysql> SELECT EXTRACT(MICROSECOND
    ->                FROM '2003-01-02 10:30:00.000123');
        -> 123
```

Comment (RingsC) [reply mysql-test/suite/tianmu/t/issue845.test:18]: we should be aware of corner cases. 

[EXTRACT(unit FROM **date**)](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_extract)

```
Returns NULL if date is NULL.
```

File: storage/tianmu/core/mysql_expression.cpp [hunk 3]
```diff
@@ -608,6 +608,9 @@ bool operator==(Item const &l_, Item const &r_) {
         same = !std::strcmp(l->func_name(), r->func_name());
```
Comment (RingsC) [inline storage/tianmu/core/mysql_expression.cpp:609]: refine the input params using google coding style. the input params or temp vars should not be with `_`.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: mysql-test/suite/tianmu/t/issue845.test (hunk 2 above)
Comment (DandreChen) [reply mysql-test/suite/tianmu/t/issue845.test:18]: add cases to  contain these conditions

File: storage/tianmu/core/mysql_expression.cpp (hunk 3 above)
Comment (DandreChen) [reply storage/tianmu/core/mysql_expression.cpp:609]: fixed it

**Review by RingsC — APPROVED** [review]
LGTM
